### ISSUE TITLE
Section "section-data" removed after bad request to "rest/*/V1/guest-carts/*/payment-information"

### ISSUE THREAD

**Issue by archonkulis** [magento/magento2#12841]
### Preconditions

1. Magento 2.1.6 
2. php7.0-fpm

### Steps to reproduce

1. Install some 3rd party payment system ( or somehow manually fire that request - I think M2 internal devs will know what to do if you will read below ), which in the end makes a request to` rest/*/V1/guest-carts/*/payment-information` ( PayPal isn't being affected, because it makes request to set-payment-information and for that action there are no sections defined )
2. Go to payment step ( **not logged in** )
3. Enter incorrect card details ( so that action` rest/*/V1/guest-carts/*/payment-information` returns `400 Bad Request` with an error message  )
4. Reload page

### Expected result

1. `checkout-data` shouldn't be removed from local storage because it contains address data ( `shippingAddressFromData` ) and the entered address should be seen.

### Actual result

1. https://prnt.sc/hqg4hx - Because `checkout-data` has been removed from local storage after the request to ` rest/*/V1/guest-carts/*/payment-information`, after page reload, the address is corrupted, because it is simply empty ( it takes that address data from `checkout-data` section )

### Comments

As far as I understand, the reason why `checkout-data` section needs to be invalidated after the request to that action (  `rest/*/V1/guest-carts/*/payment-information` ) is that theoretically after that action there should be an order placement and after an order is placed, we don't need the `checkout-data` anymore. But, you didn't think about the requests that return some error messages in it and order isn't placed successfully and customer stays on the same page.

`checkout-data` section for that action is defined in: `/module-checkout/etc/frontend/sections.xml` under 
`<action name="rest/*/V1/guest-carts/*/payment-information">`

That 3rd party extension extends:

`extends \Magento\Checkout\Model\GuestPaymentInformationManagement`

and places an order like:

```
        try {
            $orderId = $this->cartManagement->placeOrder($cartId);
        } catch (\Magento\Framework\Exception\LocalizedException $e) {
            throw new CouldNotSaveException(
                __($e->getMessage()),
                $e
            );
        }
```

It seems that it enters the catch and throws an error and that's why there is a 400 bad request.

So, to sum this up, I think M2 should check if the requests for an action have status 200 ( e.g. order placed successfully ) and only, in that case, invalidate the sections. Or maybe you have some better ideas, not sure...

I can also provide you with some real-world example of this if you want, but only in PM ( if there is such thing ), so let me know if you would like that.

Let me know what you think.

Cheers

**Comment by magento-engcom-team** [comment]
@archonkulis , thank you for your report.
Please try to reproduce this issue on a clean installation without any third-party extensions.

**Comment by archonkulis** [comment]
@magento-engcom-team why? That extension simply uses Magento functions, it's not overriding anything. Please be more careful when reading the description. If some M2 dev would read it he wouldn't have a questions like that, so please forward this bug straight to the dev. As I already mentioned - this has nothing to do with 3rd party extensions code, but rather how it uses the Magento core functions.

**Comment by magento-engcom-team** [comment]
@archonkulis, thank you for your report.
We've created internal ticket(s) MAGETWO-82589 to track progress on the issue.

**Comment by krzksz** [comment]
This issue seems to be no longer reproducible on `2.3-develop`. I've overwritten `Magento_Checkout/js/model/place-order.js` using Chrome Overrides and replaced:
```javascript
return storage.post(
    serviceUrl, JSON.stringify(payload)
).fail(
```
with:
```javascript
return storage.post(
    serviceUrl, {}
).fail(
```
which makes the request `/guest-carts/*/payment-information` always return empty `400` response in which case error is properly displayed and no information is being lost.
